In [ ]:
import pandas as pd
import numpy as np

In [ ]:
data = pd.read_csv('heart.csv')

In [ ]:
y = data["output"]
x = data.drop("output", axis=1)

from sklearn.linear_model import LogisticRegression

logistic = LogisticRegression().fit(x, y)
logistic.score(x, y)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


0.8547854785478548

In [ ]:
x_train, y_train = x[:200], y[:200]

In [ ]:
x_train.shape, y_train.shape

((200, 13), (200,))

In [ ]:
x_test, y_test = x[200:], y[200:]
x_test.shape, y_test.shape

((103, 13), (103,))

In [ ]:
logistic_test = LogisticRegression().fit(x_train, y_train)
logistic_test.score(x_train, y_train), logistic_test.score(x_test, y_test)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


(0.9, 0.5436893203883495)

In [ ]:
N = int(x.shape[0] * 0.8)

x_train, y_train, x_test, y_test = x[:N], y[:N], x[N:], y[N:]
x_train.shape, y_train.shape, x_test.shape, y_test.shape

((242, 13), (242,), (61, 13), (61,))

In [ ]:
logistic_test = LogisticRegression().fit(x_train, y_train)
logistic_test.score(x_train, y_train), logistic_test.score(x_test, y_test)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


(0.8884297520661157, 0.6229508196721312)

In [ ]:
data.head()

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0   63    1   3     145   233    1        0       150     0      2.3    0   
1   37    1   2     130   250    0        1       187     0      3.5    0   
2   41    0   1     130   204    0        0       172     0      1.4    2   
3   56    1   1     120   236    0        1       178     0      0.8    2   
4   57    0   0     120   354    0        1       163     1      0.6    2   

   caa  thall  output  
0    0      1       1  
1    0      2       1  
2    0      2       1  
3    0      2       1  
4    0      2       1

In [ ]:
data.tail()

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
298   57    0   0     140   241    0        1       123     1      0.2    1   
299   45    1   3     110   264    0        1       132     0      1.2    1   
300   68    1   0     144   193    1        1       141     0      3.4    1   
301   57    1   0     130   131    0        1       115     1      1.2    1   
302   57    0   1     130   236    0        0       174     0      0.0    1   

     caa  thall  output  
298    0      3       0  
299    0      3       0  
300    2      3       0  
301    1      3       0  
302    1      2       0

In [ ]:
mask = np.array([True] * N + [False] * (y.shape[0] - N))

In [ ]:
from numpy.random import shuffle

shuffle(mask)
mask

array([ True,  True, False,  True,  True,  True, False,  True,  True,
        True,  True,  True, False,  True,  True,  True,  True, False,
        True,  True,  True,  True, False,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True, False,  True,  True,
       False,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True, False,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True, False,  True,  True,  True,
       False,  True,  True,  True, False,  True,  True,  True, False,
       False, False, False,  True,  True,  True, False,  True,  True,
       False,  True,  True, False,  True,  True,  True,  True,  True,
        True, False,  True,  True,  True, False,  True, False,  True,
        True, False,

In [ ]:
x_train = x[mask]
x_train.shape

(242, 13)

In [ ]:
x_train, y_train, x_test, y_test = x[mask], y[mask], x[~mask], y[~mask]
x_train.shape, y_train.shape, x_test.shape, y_test.shape

((242, 13), (242,), (61, 13), (61,))

In [ ]:
logistic_test = LogisticRegression().fit(x_train, y_train)
logistic_test.score(x_train, y_train), logistic_test.score(x_test, y_test)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


(0.8677685950413223, 0.8032786885245902)

In [ ]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x, y, train_size=0.8)
x_train.shape, y_train.shape, x_test.shape, y_test.shape

((242, 13), (242,), (61, 13), (61,))

In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix, classification_report

In [ ]:
y_test_pred = logistic_test.predict(x_test)
y_train_pred = logistic_test.predict(x_train)

In [ ]:
confusion_matrix(y_train, y_train_pred)

array([[ 83,  26],
       [ 10, 123]])

In [ ]:
confusion_matrix(y_test, y_test_pred)

array([[22,  7],
       [ 1, 31]])

In [ ]:
print(classification_report(y_test, y_test_pred))

              precision    recall  f1-score   support

           0       0.96      0.76      0.85        29
           1       0.82      0.97      0.89        32

    accuracy                           0.87        61
   macro avg       0.89      0.86      0.87        61
weighted avg       0.88      0.87      0.87        61



In [ ]:
metrics = pd.DataFrame({
    "Train": [
        accuracy_score(y_train, y_train_pred),
        precision_score(y_train, y_train_pred),
        recall_score(y_train, y_train_pred),
        f1_score(y_train, y_train_pred),
    ],
    "Test": [
        accuracy_score(y_test, y_test_pred),
        precision_score(y_test, y_test_pred),
        recall_score(y_test, y_test_pred),
        f1_score(y_test, y_test_pred),
    ],
}, index = ["Accuracy", "Precision", "Recall", "F1"])

metrics

Train      Test
Accuracy   0.851240  0.868852
Precision  0.825503  0.815789
Recall     0.924812  0.968750
F1         0.872340  0.885714

# **1**

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import Pipeline
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB

data = pd.read_csv('heart.csv')
y = data["output"]
X = data.drop("output", axis=1)

# Разделяем данные на обучающий и тестовый наборы
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Инициализируем модели
models = {
    "Logistic Regression": LogisticRegression(),
    "Polynomial Features": Pipeline([
        ("poly_features", PolynomialFeatures(degree=2)),
        ("log_reg", LogisticRegression())
    ]),
    "Support Vector Machine": SVC(),
    "Multi-layer Perceptron": MLPClassifier(),
    "Decision Tree": DecisionTreeClassifier(),
    "K Nearest Neighbors": KNeighborsClassifier(),
    "Naive Bayes": GaussianNB()
}

# Создаем пустой list для хранения результатов
results = []

# Обучаем модели и оцениваем их эффективность
for name, model in models.items():
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)

    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred)
    recall = recall_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred)

    # Добавляем результаты
    results.append({
        "Model": name,
        "Accuracy": accuracy,
        "Precision": precision,
        "Recall": recall,
        "F1 Score": f1
    })

results_df = pd.DataFrame(results)
print(results_df)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _c

                    Model  Accuracy  Precision   Recall  F1 Score
0     Logistic Regression  0.885246   0.878788  0.90625  0.892308
1     Polynomial Features  0.836066   0.843750  0.84375  0.843750
2  Support Vector Machine  0.704918   0.666667  0.87500  0.756757
3  Multi-layer Perceptron  0.885246   0.878788  0.90625  0.892308
4           Decision Tree  0.836066   0.923077  0.75000  0.827586
5     K Nearest Neighbors  0.688525   0.685714  0.75000  0.716418
6             Naive Bayes  0.868852   0.900000  0.84375  0.870968


# **2**

In [ ]:
from sklearn.datasets import make_classification

X, y = make_classification(n_samples=1000, n_features=20, n_informative=5, n_redundant=0, n_classes=5, random_state=42)

In [ ]:
np.unique(y)

array([0, 1, 2, 3, 4])

In [ ]:
# Разделяем данные на обучающий и тестовый наборы
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Инициализируем модели
models = {
    "Logistic Regression": LogisticRegression(),
    "Polynomial Features": Pipeline([
        ("poly_features", PolynomialFeatures(degree=2)),
        ("log_reg", LogisticRegression())
    ]),
    "Support Vector Machine": SVC(),
    "Multi-layer Perceptron": MLPClassifier(),
    "Decision Tree": DecisionTreeClassifier(),
    "K Nearest Neighbors": KNeighborsClassifier(),
    "Naive Bayes": GaussianNB()
}

# Создаем пустой list для хранения результатов
results = []

# Обучаем модели и оцениваем их эффективность
for name, model in models.items():
    print(name)
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)

    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred, average='weighted')
    recall = recall_score(y_test, y_pred, average='weighted')
    f1 = f1_score(y_test, y_pred, average='weighted')

    # Добавляем результаты
    results.append({
        "Model": name,
        "Accuracy": accuracy,
        "Precision": precision,
        "Recall": recall,
        "F1 Score": f1
    })

results_df = pd.DataFrame(results)
print(results_df)

Logistic Regression
Polynomial Features
Support Vector Machine


/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Multi-layer Perceptron


/usr/local/lib/python3.10/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Decision Tree
K Nearest Neighbors
Naive Bayes
                    Model  Accuracy  Precision  Recall  F1 Score
0     Logistic Regression     0.520   0.508455   0.520  0.509534
1     Polynomial Features     0.580   0.587585   0.580  0.581990
2  Support Vector Machine     0.695   0.714901   0.695  0.691878
3  Multi-layer Perceptron     0.690   0.706202   0.690  0.690073
4           Decision Tree     0.580   0.583326   0.580  0.576877
5     K Nearest Neighbors     0.560   0.582201   0.560  0.560786
6             Naive Bayes     0.540   0.540672   0.540  0.523846


In [ ]:
# Разделяем данные на обучающий и тестовый наборы
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Инициализируем модели
models = {
    "Logistic Regression": LogisticRegression(),
    "Polynomial Features": Pipeline([
        ("poly_features", PolynomialFeatures(degree=2)),
        ("log_reg", LogisticRegression())
    ]),
    "Support Vector Machine": SVC(),
    "Multi-layer Perceptron": MLPClassifier(),
    "Decision Tree": DecisionTreeClassifier(),
    "K Nearest Neighbors": KNeighborsClassifier(),
    "Naive Bayes": GaussianNB()
}

# Создаем пустой list для хранения результатов
results = []

# Обучаем модели и оцениваем их эффективность
for name, model in models.items():
    print(name)
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)

    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred, average='macro')
    recall = recall_score(y_test, y_pred, average='macro')
    f1 = f1_score(y_test, y_pred, average='macro')

    # Добавляем результаты
    results.append({
        "Model": name,
        "Accuracy": accuracy,
        "Precision": precision,
        "Recall": recall,
        "F1 Score": f1
    })

results_df = pd.DataFrame(results)
print(results_df)

Logistic Regression
Polynomial Features


/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Support Vector Machine
Multi-layer Perceptron
Decision Tree
K Nearest Neighbors
Naive Bayes
                    Model  Accuracy  Precision    Recall  F1 Score
0     Logistic Regression     0.520   0.510053  0.524368  0.512702
1     Polynomial Features     0.580   0.587432  0.581613  0.582743
2  Support Vector Machine     0.695   0.712664  0.699434  0.693557
3  Multi-layer Perceptron     0.625   0.627071  0.629198  0.621621
4           Decision Tree     0.580   0.592237  0.580382  0.580273
5     K Nearest Neighbors     0.560   0.583307  0.563044  0.562726
6             Naive Bayes     0.540   0.541256  0.545614  0.527274


/usr/local/lib/python3.10/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


# **Параметр average в метриках**

In [ ]:
from sklearn.metrics import precision_score, recall_score, f1_score

# Пример метрик для бинарной классификации
y_true_binary = [0, 1, 1, 0]
y_pred_binary = [0, 1, 0, 1]

precision_binary = precision_score(y_true_binary, y_pred_binary, average='binary')
recall_binary = recall_score(y_true_binary, y_pred_binary, average='binary')
f1_binary = f1_score(y_true_binary, y_pred_binary, average='binary')

print("Binary Precision:", precision_binary)
print("Binary Recall:", recall_binary)
print("Binary F1 Score:", f1_binary)

# Пример метрик для многоклассовой классификации
y_true_multiclass = [0, 1, 2, 0, 1, 2]
y_pred_multiclass = [0, 2, 1, 0, 0, 1]

precision_micro = precision_score(y_true_multiclass, y_pred_multiclass, average='micro')
precision_macro = precision_score(y_true_multiclass, y_pred_multiclass, average='macro')
precision_weighted = precision_score(y_true_multiclass, y_pred_multiclass, average='weighted')

print("Micro Precision:", precision_micro)
print("Macro Precision:", precision_macro)
print("Weighted Precision:", precision_weighted)


Binary Precision: 0.5
Binary Recall: 0.5
Binary F1 Score: 0.5
Micro Precision: 0.3333333333333333
Macro Precision: 0.2222222222222222
Weighted Precision: 0.2222222222222222


1. **'binary':** Этот параметр подходит для бинарной классификации.

2. **'micro':** Этот параметр вычисляет метрику глобально, путем подсчета количества истинно положительных, ложно положительных и ложно отрицательных для всех классов, а затем вычисления метрики из этих общих значений.

3. **'macro':** Этот параметр вычисляет метрику для каждого класса независимо, а затем усредняет результаты без учета дисбаланса классов.

4. **'weighted':** Этот параметр вычисляет метрику для каждого класса независимо и взвешивает результаты по количеству объектов каждого класса.

**Precision:**
Class 0:  TP = 2,   FP = 1

Class 1:  TP = 0,   FP = 2

Class 2:  TP = 0,   FP = 1


**Micro precision** = (TP_0 + TP_1 + TP_2) / (TP_0 + TP_1 + TP_2 + FP_0 + FP_1 + FP_2) = (2 + 0 + 0) / (2 + 0 + 0 + 1 + 2 + 1) = 0.3333

**Macro precision:**
Precision_0 = TP_0 / (TP_0 + FP_0) = 2 / 3

Precision_1 = TP_1 / (TP_1 + FP_1) = 0

Precision_2 = TP_2 / (TP_2 + FP_2) = 0

Macro precision = (Precision_0 + Precision_1 + Precision_2) / 3

**Weighted precision:**
Class 0: встречается 2 раза

Class 1: встречается 2 раза

Class 2: встречается 2 раза

Таким образом, каждый класс имеет одинаковый вес, и взвешенная precision будет равна средней precision по всем классам, причем каждый класс учитывается пропорционально его частоте.


Weighted precision = (2 * Precision_0 + 2 * Precision_1 + 2 * Precision_2) / 6

In [ ]:
# Пример многоклассовой классификации с дисбалансом классов
y_true_multiclass = [0, 1, 1, 0, 1, 2, 2, 2, 2, 2]  # Целевая переменная
y_pred_multiclass = [0, 0, 1, 0, 1, 1, 2, 2, 0, 0]  # Предсказанные значения

# Вычисляем взвешенную precision
precision_weighted = precision_score(y_true_multiclass, y_pred_multiclass, average='weighted')

print("Weighted Precision:", precision_weighted)

Weighted Precision: 0.78


Weighted precision:

Class 0: встречается 2 раза

Class 1: встречается 3 раза

Class 2: встречается 5 раз

Class 0:  TP = 2,   FP = 3

Class 1:  TP = 2,   FP = 1

Class 2:  TP = 2,   FP = 0

Precision_0 = TP_0 / (TP_0 + FP_0) = 2 / 5

Precision_1 = TP_1 / (TP_1 + FP_1) = 2 / 3

Precision_2 = TP_2 / (TP_2 + FP_2) = 1

Weighted precision = (2 * Precision_0 + 3 * Precision_1 + 5 * Precision_2) / 10 = 0.78

1. **Параметр 'binary':** использовать для бинарной классификации.
2. **Параметр 'micro':** использовать при дисбалансе классов для оценки общей производительности модели.
3. **Параметр 'macro':** использовать для оценки средней производительности модели для каждого класса при равнозначности классов.
4. **Параметр 'weighted':** использовать при дисбалансе классов для оценки производительности модели с учетом весов классов.

# **3**

In [ ]:
import numpy as np
import pandas as pd
from sklearn.datasets import make_regression
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.svm import SVR
from sklearn.neural_network import MLPRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import mean_squared_error, mean_absolute_error, mean_absolute_percentage_error, r2_score, mean_squared_log_error, median_absolute_error, max_error
from sklearn.pipeline import Pipeline

# Генерируем синтетические данные для регрессии
X, y = make_regression(n_samples=1000, n_features=10, noise=0.1, random_state=42)

# Разделяем данные на обучающий и тестовый наборы
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Инициализируем модели регрессии
models = {
    "Linear Regression": LinearRegression(),
    "Polynomial Regression": Pipeline([
        ("poly_features", PolynomialFeatures(degree=2)),
        ("lin_reg", LinearRegression())
    ]),
    "Support Vector Regression": SVR(),
    "Multi-layer Perceptron Regression": MLPRegressor(),
    "Decision Tree Regression": DecisionTreeRegressor(),
    "K Nearest Neighbors Regression": KNeighborsRegressor()
}

# Создаем пустой list для хранения результатов
results = []

# Обучаем модели и оцениваем их эффективность
for name, model in models.items():
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)

    mse = mean_squared_error(y_test, y_pred)
    rmse = np.sqrt(mse)
    mae = mean_absolute_error(y_test, y_pred)
    mape = mean_absolute_percentage_error(y_test, y_pred)
    r2 = r2_score(y_test, y_pred)
    # msle = mean_squared_log_error(y_test, y_pred)
    median_absolute = median_absolute_error(y_test, y_pred)
    max_err = max_error(y_test, y_pred)

    # Добавляем результаты
    results.append({
        "Model": name,
        "MSE": mse,
        "RMSE": rmse,
        "MAE": mae,
        "MAPE": mape,
        "R2": r2,
        # "MSLE": msle,
        "Median Absolute Error": median_absolute,
        "Max Error": max_err
    })

results_df = pd.DataFrame(results)
print(results_df)

                               Model           MSE        RMSE        MAE  \
0                  Linear Regression      0.009512    0.097529   0.077733   
1              Polynomial Regression      0.009598    0.097970   0.077568   
2          Support Vector Regression  12758.146130  112.951964  89.705079   
3  Multi-layer Perceptron Regression    604.348804   24.583507  17.525778   
4           Decision Tree Regression   6540.899783   80.875829  63.140214   
5     K Nearest Neighbors Regression   3728.344294   61.060169  48.860734   

       MAPE        R2  Median Absolute Error   Max Error  
0  0.002136  0.999999               0.064845    0.298754  
1  0.002160  0.999999               0.063760    0.304303  
2  0.858361  0.243788              72.353132  318.240244  
3  0.434575  0.964179              12.784675   89.779499  
4  2.049088  0.612302              47.074819  212.420084  
5  1.290987  0.779010              41.568213  209.070515  


/usr/local/lib/python3.10/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
